osu.Game/Beatmaps/WorkingBeatmap.cs

```diff
@@ -108,7 +106,16 @@ public void CancelAsyncLoad()
 
         public virtual bool TrackLoaded => track != null;
 
-        public Track LoadTrack() => track = GetBeatmapTrack() ?? GetVirtualTrack(1000);
+        public Track LoadTrack()
+        {
+            track = GetBeatmapTrack() ?? GetVirtualTrack(1000);
+
+            // the track may have changed, recycle the current waveform.
+            waveform?.Dispose();
+            waveform = null;
```

Comment:
I'm not 100% sure how safe it is to dispose a waveform from `WorkingBeatmap` itself, but nothing failed in game so it seems fine. There are no unmanaged/BASS resources to be freed by the disposal method, but I'm concerned about the GC handle causing the callbacks to be leaked on iOS, so I've pushed it.

Reply:
Seems okay to go with until it fails.